File: doc/blog/thread/sql-trace.rb
# sql-trace.rb
require 'sqlite3'
require './trace.rb'
class Main
  def main
    db = SQLite3::Database.new(':memory:');
    thread = Thread.new {
      x = trace(db, :execute, 'SELECT 2 * 3')
      trace(self, :puts, x)
    }
    x = trace(db, :execute, 'SELECT 4 * 5')
    trace(self, :puts, x)
    thread.join()
    return 0
  end
end
trace(Main.new, :main)

File: doc/blog/thread/sql-trace.mjs
// sql-trace.mjs
import util from "util";
import sqlite3 from "sqlite3";
import { trace, traceAsync } from "./trace.mjs";
const db = new sqlite3.Database(":memory:");
const execute = util.promisify(db.all.bind(db));
const logTrace = (x) => trace(log, x);
const main = () => {
  traceAsync(execute, "SELECT 2 * 3 as x").then(([{ x }]) => {
    trace(console.log, x);
  });
  traceAsync(execute, "SELECT 4 * 5 as x").then(([{ x }]) => {
    trace(console.log, x);
  });
  return 0;
};
trace(main);
